File: windows/src/DmxMoney.ViewModels/Navigation.cs
using DmxMoney.Interop;

namespace DmxMoney.ViewModels;

public enum AppRoute
{
    Dashboard,
    Accounts,
    Transactions,
    Budget,
    Scheduled,
    Analytics,
    Predictions,
    Categories,
    Settings,
}

public sealed record SidebarSection(string Title, AppRoute[] Routes);

public static class AppRoutes
{
    public static readonly AppRoute[] All = Enum.GetValues<AppRoute>();

    public static readonly SidebarSection[] SidebarSections =
    [
        new("Général", [AppRoute.Dashboard, AppRoute.Accounts, AppRoute.Transactions]),
        new("Finances", [AppRoute.Budget, AppRoute.Scheduled]),
        new("Analyses", [AppRoute.Analytics, AppRoute.Predictions]),
    ];

    public static readonly AppRoute[] FooterRoutes = [AppRoute.Categories, AppRoute.Settings];

    public static string Title(this AppRoute route) => route switch
    {
        AppRoute.Dashboard => "Vue d'ensemble",
        AppRoute.Accounts => "Mes Comptes",
        AppRoute.Transactions => "Journal",
        AppRoute.Budget => "Budget",
        AppRoute.Scheduled => "Échéancier",
        AppRoute.Analytics => "Analyses",
        AppRoute.Predictions => "Prédictions",
        AppRoute.Categories => "Catégories",
        _ => "Paramètres",
    };

    /// <summary>Nom de l'icône Lucide, identique aux autres plateformes.</summary>
    public static string Icon(this AppRoute route) => route switch
    {
        AppRoute.Dashboard => "LayoutDashboard",
        AppRoute.Accounts => "Wallet",
        AppRoute.Transactions => "Receipt",
        AppRoute.Budget => "Calculator",
        AppRoute.Scheduled => "CalendarClock",
        AppRoute.Analytics => "PieChart",
        AppRoute.Predictions => "TrendingUp",
        AppRoute.Categories => "Tag",
        _ => "Settings",
    };

    public static bool UsesAccountFilter(this AppRoute route) => route is AppRoute.Dashboard or AppRoute.Transactions
        or AppRoute.Budget or AppRoute.Scheduled or AppRoute.Analytics or AppRoute.Predictions;

    public static bool ShowsBalances(this AppRoute route) => route is AppRoute.Dashboard or AppRoute.Transactions;
}

/// <summary>Formulaires demandés par les pages ; l'hôte les présente en boîte de dialogue.</summary>
public abstract record FormRequest
{
    public sealed record AccountForm(string? Id) : FormRequest;

    public sealed record AccountGroups : FormRequest;

    public sealed record TransactionForm(string? Id) : FormRequest;

    public sealed record CategoryForm(string? Id) : FormRequest;

    public sealed record BudgetForm(string? Id, string? CategoryId = null) : FormRequest;

    public sealed record ScheduledForm(string? Id) : FormRequest;

    public sealed record FakeTransactionForm(string? Id) : FormRequest;

    public sealed record BudgetSuggestions : FormRequest;

    public sealed record ScheduledSuggestions : FormRequest;

    public sealed record RestoreBackup(string Content, string FileName) : FormRequest;

    public sealed record StatementImport(string Content, string FileName) : FormRequest;

    public sealed record WhatsNew : FormRequest;
}

/// <summary>Choix du type d'opération dans un formulaire.</summary>
public sealed record KindOption(TransactionType Value, string Label);

/// <summary>Choix d'une fréquence dans un formulaire.</summary>
public sealed record PeriodicityOption(Periodicity Value, string Label);

public sealed record ConfirmRequest(string Title, string Message, string ConfirmTitle, bool Destructive, Action Action);

/// <summary>Services propres à la plateforme (sélecteurs de fichiers, presse-papiers, mises à jour).</summary>
public interface IPlatformServices
{
    Task ExportBackupAsync(string content, string suggestedFileName);

    Task<(string Content, string FileName)?> PickImportFileAsync();

    void CopyToClipboard(string text);

    bool UpdateAvailable { get; }

    Task CheckForUpdatesAsync();
}

public sealed class NullPlatformServices : IPlatformServices
{
    public bool UpdateAvailable => false;

    public Task ExportBackupAsync(string content, string suggestedFileName) => Task.CompletedTask;

    public Task<(string Content, string FileName)?> PickImportFileAsync() => Task.FromResult<(string, string)?>(null);

    public void CopyToClipboard(string text)
    {
    }

    public Task CheckForUpdatesAsync() => Task.CompletedTask;
}

File: apple/Packages/DmxKit/Sources/DmxKit/Store/Navigation.swift
import Foundation

/// Pages de l'application, identiques à la barre latérale de DmxMoney 1.x.
public enum AppRoute: String, CaseIterable, Identifiable {
    case dashboard
    case accounts
    case transactions
    case budget
    case scheduled
    case analytics
    case predictions
    case categories
    case settings

    public var id: String { rawValue }

    public var title: String {
        switch self {
        case .dashboard: return "Vue d'ensemble"
        case .accounts: return "Mes Comptes"
        case .transactions: return "Journal"
        case .budget: return "Budget"
        case .scheduled: return "Échéancier"
        case .analytics: return "Analyses"
        case .predictions: return "Prédictions"
        case .categories: return "Catégories"
        case .settings: return "Paramètres"
        }
    }

    /// Libellé court de la barre d'onglets mobile.
    public var shortTitle: String {
        switch self {
        case .dashboard: return "Accueil"
        case .accounts: return "Comptes"
        default: return title
        }
    }

    public var icon: String {
        switch self {
        case .dashboard: return "LayoutDashboard"
        case .accounts: return "Wallet"
        case .transactions: return "Receipt"
        case .budget: return "Calculator"
        case .scheduled: return "CalendarClock"
        case .analytics: return "PieChart"
        case .predictions: return "TrendingUp"
        case .categories: return "Tag"
        case .settings: return "Settings"
        }
    }

    /// Pages où le filtre global de comptes s'applique.
    public var usesAccountFilter: Bool {
        switch self {
        case .dashboard, .transactions, .budget, .analytics, .predictions, .scheduled: return true
        case .accounts, .categories, .settings: return false
        }
    }

    /// Pages qui affichent les soldes Pointé / Actuel dans la barre d'outils.
    public var showsBalances: Bool {
        true
    }

    public static let sidebarSections: [SidebarSection] = [
        SidebarSection(title: "Général", routes: [.dashboard, .accounts, .transactions]),
        SidebarSection(title: "Finances", routes: [.budget, .scheduled]),
        SidebarSection(title: "Analyses", routes: [.analytics, .predictions]),
    ]

    public static let footerRoutes: [AppRoute] = [.categories, .settings]
}

public struct SidebarSection: Identifiable {
    public let title: String
    public let routes: [AppRoute]
    public var id: String { title }
}

/// Formulaires présentés par l'hôte (feuille AppKit sur macOS, sheet SwiftUI sur iOS).
public enum FormRequest: Identifiable, Equatable {
    case account(id: String?)
    case accountGroups
    case transaction(id: String?)
    case category(id: String?)
    case budget(id: String?)
    case scheduled(id: String?)
    case fakeTransaction(id: String?)
    case restoreBackup(content: String, fileName: String)
    case statementImport(content: String, fileName: String)
    case whatsNew
    case newBudget(categoryId: String)
    case budgetSuggestions
    case scheduledSuggestions

    public var id: String {
        switch self {
        case let .account(id): return "account-\(id ?? "new")"
        case .accountGroups: return "account-groups"
        case let .transaction(id): return "transaction-\(id ?? "new")"
        case let .category(id): return "category-\(id ?? "new")"
        case let .budget(id): return "budget-\(id ?? "new")"
        case let .scheduled(id): return "scheduled-\(id ?? "new")"
        case let .fakeTransaction(id): return "fake-\(id ?? "new")"
        case let .restoreBackup(_, fileName): return "restore-\(fileName)"
        case let .statementImport(_, fileName): return "import-\(fileName)"
        case .whatsNew: return "whats-new"
        case let .newBudget(categoryId): return "budget-for-\(categoryId)"
        case .budgetSuggestions: return "budget-suggestions"
        case .scheduledSuggestions: return "scheduled-suggestions"
        }
    }

    /// Largeur préférée de la feuille sur macOS et iPad.
    public var preferredWidth: Double {
        switch self {
        case .statementImport, .budgetSuggestions, .scheduledSuggestions: return 640
        case .category, .account: return 520
        case .whatsNew: return 480
        default: return 460
        }
    }
}

/// Demande de confirmation (suppression, restauration…).
public struct ConfirmRequest: Identifiable {
    public let id = UUID()
    public let title: String
    public let message: String
    public let confirmTitle: String
    public let destructive: Bool
    public let action: () -> Void
}
